src/main/java/com/maxmind/db/Decoder.java

```diff
@@ -220,65 +207,257 @@ static int decodeInteger(ByteBuffer buffer, int base, int size) {
         }
-        return new DoubleNode(this.buffer.getDouble());
+        return this.buffer.getDouble();
     }
 
-    private FloatNode decodeFloat(int size) throws InvalidDatabaseException {
+    private float decodeFloat(int size) throws InvalidDatabaseException {
         if (size != 4) {
             throw new InvalidDatabaseException(
                     "The MaxMind DB file's data section contains bad data: "
                             + "invalid size of float.");
         }
-        return new FloatNode(this.buffer.getFloat());
+        return this.buffer.getFloat();
     }
 
-    private static BooleanNode decodeBoolean(int size)
+    private static boolean decodeBoolean(int size)
             throws InvalidDatabaseException {
         switch (size) {
             case 0:
-                return BooleanNode.FALSE;
+                return false;
             case 1:
-                return BooleanNode.TRUE;
+                return true;
             default:
                 throw new InvalidDatabaseException(
                         "The MaxMind DB file's data section contains bad data: "
                                 + "invalid size of boolean.");
         }
     }
 
-    private JsonNode decodeArray(int size) throws IOException {
+    private <T> Object decodeArray(int size, Class<T> cls)
+            throws IOException,
+                   InstantiationException,
+                   IllegalAccessException,
+                   InvocationTargetException,
+                   DeserializationException {
+        // If we're decoding into a Map, the second level is an Object.class.
+        // See Object.class in decodeMapIntoMap().
+        if (cls.equals(List.class) || cls.equals(Object.class)) {
+            ArrayList<Object> array = new ArrayList<>(size);
+
+            for (int i = 0; i < size; i++) {
+                Object e = this.decode(Object.class);
+                array.add(e);
+            }
+
+            return array;
+        }
+
+        if (!cls.isArray()) {
+            throw new DeserializationException();
+        }
+
+        Class<?> elementClass = cls.getComponentType();
+
+        Object array = Array.newInstance(elementClass, size);
 
-        List<JsonNode> array = new ArrayList<>(size);
         for (int i = 0; i < size; i++) {
-            JsonNode r = this.decode();
-            array.add(r);
+            Object e = this.decode(elementClass);
+            Array.set(array, i, e);
+        }
+
+        return array;
+    }
+
+    private <T> Object decodeMap(int size, Class<T> cls)
+            throws IOException,
+                   InstantiationException,
+                   IllegalAccessException,
+                   InvocationTargetException {
+        // We look for Object.class because if we're decoding a nested map into
+        // a Map, the second level is an Object.class. See Object.class in
+        // decodeMapIntoMap().
+        if (cls.equals(Map.class) || cls.equals(Object.class)) {
```

Comment:
Couldn't we support anything that is assignable to `Map`? It looks like `geoip2` does just use `Map` so maybe that could be a future enhancement if needed.

Reply:
Yeah I think so. I made a change that I think will do that!